### PR TITLE
[firebase-recaptcha] Add missing _reset method to modal verifier

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This adds the missing (internal) `_reset` method to the Recaptcha Model verifier.

See issues:
  - https://github.com/firebase/firebase-js-sdk/issues/5688
  - #14780

# How

- Added additional comments referring to the APIs we need to "replace" in React Native.
- Added missing `_reset` method in Recaptcha Verifier Modal
- Removed old workarounds for previous Firebase issues in test suite
- Updated example code to work with v9 on SDK 43 and unversioned

# Test Plan

See #15117

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by yoavTaieb** [comment]
Any ETA on when it will be push to master?

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ---
 title: FirebaseRecaptcha
-sourceCodeUrl: 'https://github.com/expo/expo/tree/master/packages/expo-firebase-recaptcha'
+sourceCodeUrl: 'https://github.com/expo/expo/tree/sdk-43/packages/expo-firebase-recaptcha'
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:3]: This looks like unintended change.

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 2]
```diff
@@ -83,29 +85,32 @@ import { getAuth, PhoneAuthProvider } from 'firebase/auth';
...
   const attemptInvisibleVerification = false;
 
   return (
     <View style={{ padding: 20, marginTop: 50 }}>
       <FirebaseRecaptchaVerifierModal
         ref={recaptchaVerifier}
-        firebaseConfig={firebaseConfig}
-        attemptInvisibleVerification={attemptInvisibleVerification}
+        firebaseConfig={app.options}
+        // attemptInvisibleVerification
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:113]: Is it necessary to leave the commented out prop in here?

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 3]
```diff
@@ -83,29 +85,32 @@ import { getAuth, PhoneAuthProvider } from 'firebase/auth';
...
+  const [message, showMessage] = React.useState();
   const attemptInvisibleVerification = false;
 
   return (
     <View style={{ padding: 20, marginTop: 50 }}>
       <FirebaseRecaptchaVerifierModal
         ref={recaptchaVerifier}
-        firebaseConfig={firebaseConfig}
-        attemptInvisibleVerification={attemptInvisibleVerification}
+        firebaseConfig={app.options}
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:112]: Is there a reason why you did not have used the `firebaseConfig` variable in here?

File: packages/expo-firebase-recaptcha/CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - Update firebase imports for web to support the v9 firebase JS sdk. ([#14829](https://github.com/expo/expo/pull/14829) by [@JJK801](https://github.com/JJK801))
+- Add missing _reset method to modal verifier ([#15102](https://github.com/expo/expo/pull/15102) by [@byCedric](https://github.com/byCedric))
```
Comment (Simek) [inline packages/expo-firebase-recaptcha/CHANGELOG.md:12]:
```suggestion
- Add missing `_reset` method to modal verifier. ([#15102](https://github.com/expo/expo/pull/15102) by [@byCedric](https://github.com/byCedric))
```

File: apps/test-suite/tests/FirebaseJSSDK.js [hunk 5]
```diff
@@ -191,6 +182,20 @@ export async function test({ describe, it, expect, beforeAll }) {
...
+      it('exports signInWithPhoneNumber', () => {
+        expect(signInWithPhoneNumber).not.toBe(undefined);
+      });
+
+      // see: https://github.com/firebase/firebase-js-sdk/issues/5638
+      it('exports PhoneAuthProvider', () => {
+        expect(PhoneAuthProvider).not.toBe(undefined);
+      });
+    });
+  });
```
Comment (Simek) [inline apps/test-suite/tests/FirebaseJSSDK.js:198]:
```suggestion
  describe('regression', () => {
    // see: https://github.com/firebase/firebase-js-sdk/issues/5638
    describe('firebase/auth', () => {
      it('exports signInWithPhoneNumber', () => {
        expect(signInWithPhoneNumber).not.toBe(undefined);
      });

      it('exports PhoneAuthProvider', () => {
        expect(PhoneAuthProvider).not.toBe(undefined);
      });
    });
  });
```
If it is the same link, we can mention it only once. 🙂

**Review by Simek — APPROVED** [review]
Besides the comments above, LGTM! 👍

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md (hunk 1 above)
Comment (byCedric) [reply docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:3]: ah yes, thanks for catching this!

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md (hunk 2 above)
Comment (byCedric) [reply docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:113]: In this example, I'm mostly trying to steer people to try out both. Adding a comment was too long, but just leaving this there should be enough for people to try both (I think)

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md (hunk 3 above)
Comment (byCedric) [reply docs/pages/versions/unversioned/sdk/firebase-recaptcha.md:112]: Usually, you don't want to reuse the firebase config directly in your RN app. E.g. if you need to test, you can just mock the app and you will have a handle on this as well. Or if you have a provider, you probably don't want to leak the config to all your components using this captcha.

It's one of many ways to provide this config, I think this is the most versatile one. Where you can just initialize that app with whatever you need, and only that will be used. I can revert it though, don't have a heavy opinion on this

File: apps/test-suite/tests/FirebaseJSSDK.js (hunk 5 above)
Comment (byCedric) [reply apps/test-suite/tests/FirebaseJSSDK.js:198]: Yeah, thats true. When writing this I thought more of "1 test per issue, every issue should have an explanation". But we can remove one of them :)

**Comment by adriaanbalt** [comment]
@byCedric can this be merged?  it is blocking my users from authenticating on my latest v43 AppStore update. great work btw

**Comment by abeaclark** [comment]
Agreed, thank you for the fix, would love to see this ship.

Would also love to have this one published: https://github.com/expo/expo/pull/15038

Thanks so much!

**Comment by yoavTaieb** [comment]
Hey, any news on the merge? It's blocking my app update... Thanks :-)

**Comment by adriaanbalt** [comment]
@Simek are you able to merge this?

**Comment by adriaanbalt** [comment]
@Simek @byCedric does the npm registry also need to be updated?  what is the estimated time for that to be completed?

**Comment by adriaanbalt** [comment]
@brentvatne maybe you have some insight into when we can expect to receive this update?  im unable to authenticate users by Phone and therefore cannot migrate to SDK 43 with confidence.

**Comment by toey919** [comment]
@Simek @byCedric Can I get published version as soon as possible? I am still getting this error on expo sdk 43 on EAS build. Because I think this was not unpublished yet.

**Comment by aliexalter** [comment]
Is it published or merged. How can I get this update ? Do i need to reinstall or just run npm update ?

**Comment by aliexalter** [comment]
I run `expo install expo-firebase-recaptcha` and look at the files it did not has updated version on line 85 of *Modal.js and tsx files. I think npm registry is not updated yet. right ?

**Comment by aaronbesson** [comment]
Any update or work around?

**Comment by toey919** [comment]
> Any update or work around?
> 
> work around
> 
> edit: FirebaseRecaptchaVerifierModal.js in Expo Firebase Recaptcha/build
> 
> add:
> 
> ` async _reset(){ }`
> 
> around line 31 or so

This is ok for local environment. But `EAS build` uses EAS build servers environment, not local one.
So EAS server install dependencies and build standalone app.
How can I edit node_module during EAS build?
I think it is impossible. So I hope this issue is published soon.

**Comment by adscud** [comment]
> > Any update or work around?
> > work around
> > edit: FirebaseRecaptchaVerifierModal.js in Expo Firebase Recaptcha/build
> > add:
> > ` async _reset(){ }`
> > around line 31 or so
> 
> This is ok for local environment. But `EAS build` uses EAS build servers environment, not local one. So EAS server install dependencies and build standalone app. How can I edit node_module during EAS build? I think it is impossible. So I hope this issue is published soon.

Maybe you have to fork the project and use yours while it's not fixed.

**Comment by aliexalter** [comment]
It's the npm that is not updated so at EAS build it gets old version from npm. I think in package.json install package directly from git url if it's possible. Because code is already merged to master I don't know why it's not updated in npm yet.

**Comment by Cheesetouched** [comment]
Any updates on this? Looks like it's still not updated :/

**Comment by toey919** [comment]
This is working for me.
https://github.com/expo/expo/pull/15398#issuecomment-993413962

**Comment by vikasgurjar** [comment]
Change "expo-firebase-recaptcha": "~2.0.2" to "expo-firebase-recaptcha": "~2.1.0" in your package.json and run npm i/yarn install

**Comment by adriaanbalt** [comment]
This has started to work for me hurrah :)

**Comment by adscud** [comment]
Expo 44 comes soon, no more hack will be needed :D

**Comment by rcb4t2** [comment]
@Simek @byCedric Hey y'all, thank you for updating this! However, the `FirebaseRecaptchaVerifier` class is also missing a reset method

Please see https://github.com/expo/expo/issues/16493

**Pull request merged**